File: tuan06/cai_tui.jl
# Quy uoc chi so: Julia danh chi so mang tu 1.
# Bang f co 6 hang (i chay tu 0 den n) va 12 cot (j chay tu 0 den W).
# Do do, trang thai f[i][j] cua bai toan duoc luu tai f[i + 1, j + 1].

function cai_tui()
    n = 5
    W = 11
    ten = ["A", "B", "C", "D", "E"]
    w = [2, 3, 4, 5, 7]
    v = [3, 7, 9, 12, 16]

    f = zeros(Int, n + 1, W + 1)

    for i in 1:n
        for j in 0:W
            f[i + 1, j + 1] = f[i, j + 1]
            if w[i] <= j
                f[i + 1, j + 1] = max(f[i + 1, j + 1], f[i, j - w[i] + 1] + v[i])
            end
        end
    end

    println("Bang f (kich thuoc 6 x 12):")
    for i in 0:n
        for j in 0:W
            print(lpad(f[i + 1, j + 1], 4))
        end
        println()
    end

    chon = String[]
    j = W
    for i in n:-1:1
        if f[i + 1, j + 1] != f[i, j + 1]
            push!(chon, ten[i])
            j -= w[i]
        end
    end
    reverse!(chon)

    println("Gia tri lon nhat f[5][11]: ", f[n + 1, W + 1])
    println("Tap do vat duoc chon: ", join(chon, ", "))
end

cai_tui()

File: tuan06/cai_tui.cpp
#include <iostream>
#include <vector>
#include <string>
#include <iomanip>
#include <algorithm>

using namespace std;

int main() {
    int n = 5;
    int W = 11;
    vector<string> ten = {"A", "B", "C", "D", "E"};
    vector<int> w = {2, 3, 4, 5, 7};
    vector<int> v = {3, 7, 9, 12, 16};

    vector<vector<int>> f(n + 1, vector<int>(W + 1, 0));

    for (int i = 1; i <= n; i++) {
        for (int j = 0; j <= W; j++) {
            f[i][j] = f[i - 1][j];
            if (w[i - 1] <= j) {
                f[i][j] = max(f[i][j], f[i - 1][j - w[i - 1]] + v[i - 1]);
            }
        }
    }

    cout << "Bang f (kich thuoc 6 x 12):\n";
    for (int i = 0; i <= n; i++) {
        for (int j = 0; j <= W; j++) {
            cout << setw(4) << f[i][j];
        }
        cout << "\n";
    }

    vector<string> chon;
    int j = W;
    for (int i = n; i >= 1; i--) {
        if (f[i][j] != f[i - 1][j]) {
            chon.push_back(ten[i - 1]);
            j -= w[i - 1];
        }
    }
    reverse(chon.begin(), chon.end());

    cout << "Gia tri lon nhat f[5][11]: " << f[n][W] << "\n";
    cout << "Tap do vat duoc chon: ";
    for (int i = 0; i < (int)chon.size(); i++) {
        cout << chon[i] << (i + 1 < (int)chon.size() ? ", " : "\n");
    }

    return 0;
}